### PR TITLE
Native patch: handling file permission properly

### PR DESCRIPTION
**Pull request by meteorcloudy**
1. If no file permission is specified in the patch file, preserve the
original file permission.

2. If file permission is specified, then set it as it is.

Fixes https://github.com/bazelbuild/bazel/issues/10913

RELNOTES: Native patch can handle file permission properly

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java [hunk 1]
```diff
@@ -229,8 +237,35 @@ private static void writeFile(Path file, List<String> content) throws IOExceptio
     FileSystemUtils.writeLinesAs(file, StandardCharsets.UTF_8, content);
   }
 
+  private static boolean getReadPermission(int permission) {
+    // Parse read permission from posix file permission notation
+    return permission / 4 % 2 == 1;
```
Comment (philwo) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: We can use bit-operators here to test for the bits in the permission bitmask :)

Executable is `0b001 = 1`
Writable is `0b010 = 2`
Readable is `0b100 = 4`

So we can replace this with:

```java
return permission & 4;
```

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java [hunk 2]
```diff
@@ -229,8 +237,35 @@ private static void writeFile(Path file, List<String> content) throws IOExceptio
...
+  }
+
+  private static boolean getWritePermission(int permission) {
+    // Parse write permission from posix file permission notation
+    return permission / 2 % 2 == 1;
```
Comment (philwo) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: Similarly:
```java
return permission & 2;
```

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java [hunk 3]
```diff
@@ -229,8 +237,35 @@ private static void writeFile(Path file, List<String> content) throws IOExceptio
...
+  }
+
+  private static boolean getExecutablePermission(int permission) {
+    // Parse executable permission from posix file permission notation
+    return permission % 2 == 1;
```
Comment (philwo) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: Similarly:
```java
return permission & 1;
```

File: src/test/java/com/google/devtools/build/lib/bazel/repository/PatchUtilTest.java [hunk 4]
```diff
@@ -378,7 +408,7 @@ public void testMissingBothOldAndNewFile() throws IOException {
     Path patchFile =
         scratch.file(
             "/root/patchfile",
-            "diff --git a/foo.cc b/foo.cc",
+            "diff --git a/ b/",
```
Comment (philwo) [inline src/test/java/com/google/devtools/build/lib/bazel/repository/PatchUtilTest.java]: How is this change related?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java [hunk 5]
```diff
@@ -245,6 +280,10 @@ private static void applyPatchToFile(
       oldContent = new ArrayList<>();
     } else {
       oldContent = readFile(inputFile);
+      // Preserve old file permission if no explicit permission is set.
+      if (filePermission == -1) {
```
Comment (philwo) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: If you want to use some more exotic Java data types, you could use `Optional<BitSet>` for the filePermission type (and `BitSet` in general, rather than an `int`, to store the permissions). Then you could use an `enum` as the keys for the `BitSet` (e.g. `EXECUTABLE = 0, WRITABLE = 1, READABLE = 2`).

But using `-1` as the indicator for "no explicit permissions" is also fine, if you prefer it. ;)

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java (hunk 1 above)
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: Thanks, I'll use 
```java 
return (permission & 4) == 4
```

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java (hunk 2 above)
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: Done.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java (hunk 3 above)
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: Done

File: src/test/java/com/google/devtools/build/lib/bazel/repository/PatchUtilTest.java (hunk 4 above)
Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/PatchUtilTest.java]: So this test case is for testing an error message is thrown when no input and output file name is specified. Before, we only parse file names from `---` and `+++` lines. But now we also parses file names from `diff --git` line, therefore we should also remove them here to keep the test passing.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java (hunk 5 above)
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java]: I tried to use BitSet, but found converting a character (eg. `7`) to a BitSet also requires some steps. The code overall didn't look much cleaner, so I'll keep use `-1` and bit operations ;)

**Comment by meteorcloudy** [comment]
Thanks for the review!!
